File: src/ingest/sync.ts
/**
 * Sync orchestrator. Mirrors sync.rs::run_sync: walks the enabled providers
 * from the config and runs each ingestion source at most once per
 * invocation. The LogOmp source runs only for the raw "omp" provider entry
 * (the others are filtered views into already-ingested data, so they skip).
 * LogClaudeCode runs once for "claude-code". AdminOpenAI / AdminAnthropic
 * each fetch their admin key from the credential store; when absent, the
 * provider status is recorded as "no_admin_key" and the source is skipped.
 *
 * Admin providers are sourced by the credential store from `src/auth/store.ts`
 * (Phase 3 module). The store contract is described in local://contracts.md;
 * if Phase 3 has not landed yet, this module will fail to type-check on
 * that import — every other call here is self-consistent.
 */
import type { Database } from "bun:sqlite";
import type { Config } from "../config/types.ts";
import { byId, type SourceKind } from "../registry.ts";
import { ingestOmp } from "./omp.ts";
import { ingestClaudeCode } from "./claudeCode.ts";
import { ingestAdminOpenai } from "./adminOpenai.ts";
import { ingestAdminAnthropic } from "./adminAnthropic.ts";
import { setProviderStatus, ensureProvider } from "../db/providers.ts";
import { load } from "../auth/store.ts";

export interface SyncSourceOutcome {
  source: string;
  inserted: number;
}

async function runOne(
  db: Database,
  dataDir: string,
  kind: SourceKind,
  full: boolean,
): Promise<{ inserted: number; status: string }> {
  switch (kind) {
    case "LogOmp":
      return ingestOmp(db, full);
    case "LogClaudeCode":
      return ingestClaudeCode(db, full);
    case "AdminOpenAI": {
      const creds = load(dataDir, "openai-api");
      const apiKey = creds.find((c) => c.type === "api_key")?.key;
      if (!apiKey) {
        ensureProvider(db, "openai-api", "OpenAI (admin)", "AdminOpenAI");
        setProviderStatus(db, "openai-api", "no_admin_key");
        return { inserted: 0, status: "no_admin_key" };
      }
      return ingestAdminOpenai(db, apiKey);
    }
    case "AdminAnthropic": {
      const creds = load(dataDir, "anthropic-api");
      const apiKey = creds.find((c) => c.type === "api_key")?.key;
      if (!apiKey) {
        ensureProvider(db, "anthropic-api", "Anthropic (admin)", "AdminAnthropic");
        setProviderStatus(db, "anthropic-api", "no_admin_key");
        return { inserted: 0, status: "no_admin_key" };
      }
      return ingestAdminAnthropic(db, apiKey);
    }
  }
}

/**
 * Run every enabled provider's ingestion source. Returns per-provider
 * inserted counts in the order they ran. omp-log providers (id != "omp")
 * are skipped — they are filtered views into already-ingested data.
 */
export async function runSync(
  db: Database,
  cfg: Config,
  dataDir: string,
  full: boolean,
): Promise<SyncSourceOutcome[]> {
  const out: SyncSourceOutcome[] = [];
  let ompDone = false;
  let ccDone = false;
  for (const prov of cfg.providers) {
    if (!prov.enabled) continue;
    const def = byId(prov.id);
    if (!def) {
      setProviderStatus(db, prov.id, "unknown");
      continue;
    }
    let shouldRun = true;
    if (def.sourceKind === "LogOmp") {
      // Only the raw "omp" entry (omp_provider === null) triggers the scan.
      if (def.ompProvider !== null || ompDone) shouldRun = false;
      else ompDone = true;
    } else if (def.sourceKind === "LogClaudeCode") {
      if (ccDone) shouldRun = false;
      else ccDone = true;
    }
    if (!shouldRun) continue;
    try {
      const r = await runOne(db, dataDir, def.sourceKind, full);
      out.push({ source: prov.id, inserted: r.inserted });
      if (r.status) setProviderStatus(db, prov.id, r.status);
    } catch (e) {
      const msg = `error: ${(e as Error).message}`;
      setProviderStatus(db, prov.id, msg);
      out.push({ source: prov.id, inserted: 0 });
    }
  }
  return out;
}

File: legacy-rust/src/sync.rs
use anyhow::Result;
use std::collections::HashMap;

use crate::config::Config;
use crate::db::Db;
use crate::registry;
use crate::source::{
    AdminAnthropicSource, AdminOpenAISource, ClaudeCodeSource, OmpSource, Source,
};

pub async fn run_sync(
    db: &Db,
    cfg: &Config,
    full: bool,
) -> Result<HashMap<String, usize>> {
    let mut counts: HashMap<String, usize> = HashMap::new();

    // Phase 1: ingest raw sources. Each source is synced at most once,
    // regardless of how many providers read from it. An omp-log provider
    // with `omp_provider == None` is the "raw" entry that triggers the
    // actual directory scan; providers with `omp_provider == Some(_)`
    // are filtered views into already-ingested data and skip ingestion.
    let mut omp_done = false;
    let mut cc_done = false;
    for prov in &cfg.providers {
        if !prov.enabled {
            continue;
        }
        let Some(def) = registry::by_id(&prov.id) else {
            db.set_provider_status(&prov.id, "unknown")?;
            continue;
        };

        let should_sync = match def.source_kind {
            registry::SourceKind::LogOmp => {
                // Only the raw "omp" entry (omp_provider == None) triggers
                // the directory scan. Per-provider entries read from the
                // already-ingested rows.
                if def.omp_provider.is_none() && !omp_done {
                    omp_done = true;
                    true
                } else {
                    false
                }
            }
            registry::SourceKind::LogClaudeCode => {
                if !cc_done {
                    cc_done = true;
                    true
                } else {
                    false
                }
            }
            registry::SourceKind::AdminOpenAI | registry::SourceKind::AdminAnthropic => true,
        };

        if should_sync {
            let outcome = dispatch(db, prov, def.source_kind, full).await;
            match outcome {
                Ok(o) => {
                    counts.insert(prov.id.clone(), o.inserted);
                    if !o.status.is_empty() {
                        let _ = db.set_provider_status(&prov.id, &o.status);
                    }
                }
                Err(e) => {
                    let msg = format!("error: {e}");
                    let _ = db.set_provider_status(&prov.id, &msg);
                    counts.insert(prov.id.clone(), 0);
                }
            }
        }
    }
    Ok(counts)
}

async fn dispatch(
    db: &Db,
    prov: &crate::config::ProviderCfg,
    kind: registry::SourceKind,
    full: bool,
) -> Result<crate::source::FetchOutcome> {
    use registry::SourceKind;
    // `full` re-reads from byte 0 by resetting the sync_state offset.
    if full {
        reset_source_sync_state(db, kind);
    }
    match kind {
        SourceKind::LogOmp => OmpSource::default().fetch(db, prov).await,
        SourceKind::LogClaudeCode => ClaudeCodeSource::default().fetch(db, prov).await,
        SourceKind::AdminOpenAI => AdminOpenAISource.fetch(db, prov).await,
        SourceKind::AdminAnthropic => AdminAnthropicSource.fetch(db, prov).await,
    }
}

/// Reset the incremental sync offsets for a source so the next fetch
/// re-reads from byte 0. This is a best-effort operation — if the
/// sync_state table doesn't have matching rows, nothing happens.
fn reset_source_sync_state(db: &Db, kind: registry::SourceKind) {
    let source = match kind {
        registry::SourceKind::LogOmp => "omp",
        registry::SourceKind::LogClaudeCode => "claude-code",
        registry::SourceKind::AdminOpenAI => "openai-api",
        registry::SourceKind::AdminAnthropic => "anthropic-api",
    };
    let conn = db.conn.lock();
    let _ = conn.execute("DELETE FROM sync_state WHERE source = ?", [source]);
}